### DISCUSSION TITLE
Reading and subsampling MGF file

### DISCUSSION THREAD

**Discussion by ekopylova** [levitsky/pyteomics#49] [Q&A] [answered]
Hello,

I would like to read a ```.mgf``` file having thousands of spectra, subsample it down to a list of specific spectra and then write the new subsampled spectra to a separate ```.mgf``` file. Is it possible to use ```mgf.write()``` in a loop where I append the spectra which match my filtering? I'm looking for something like:

```
    with mgf.MGF(input_mgf_fp) as reader:
        for spectrum in reader:
            params = spectrum.get('params')
            if params['feature_id'] in feature_ids:
                mgf.write(spectrum, output = output_fp);
```

Thanks!

**Comment by mobiusklein** [comment, marked as answer]
You could do almost exactly what you have there, just wrap each individual spectrum in a list or tuple:
```python
with mgf.MGF(input_mgf_fp) as reader:
    for spectrum in reader:
        params = spectrum.get('params')
        if params['feature_id'] in feature_ids:
            mgf.write([spectrum], output_fp)
```
The first parameter of `mgf.write` is an `iterable` of spectra, so in order to match its expectations, we need to wrap single spectrum `dict` objects in something that satisfies the expectation that `next(iter(obj)) -> dict`. This happens to work because `mgf.write` doesn't close the output stream and doesn't automatically append a file header unless you tell it to.

Another option is to use a generator to abstract away the filtering logic:
```python
def producer(reader, feature_ids):
    for spectrum in reader:
        params = spectrum.get('params')
        if params['feature_id'] in feature_ids:
            yield spectrum

 with mgf.MGF(input_mgf_fp) as reader:
    spectrum_iterator = producer(reader, feature_ids)
    mgf.write(spectrum_iterator, output_fp)
```
This approach is usually more generalizable, though it does make it a little bit harder to follow because the filtering logic is no longer inline at the place where it's used.

**Reply by ekopylova** [reply]
Thank you, the minor change worked perfectly!
